File: R/species_richness.R
#
# Authors: Dan Wismer
#
# Date: June 28th, 2023
#
# Description: Generates cumulative area raster and number of species raster 
#
# Inputs:  1. Paths to 1km .tifs
#          2. Output path for species richness rasters 
#
#
# Outputs: 1. Cumulative area raster
#         2. Count of species raster
#
#===============================================================================

library(terra)
library(pbapply)

## Start timer
start_time <- Sys.time()

# Get list of tifs
ECCC_CH_LST <- list.files("Data/Output/ECCC_CH", pattern='.tif$', full.names = TRUE)
ECCC_SAR_LST <- list.files("Data/Output/ECCC_SAR", pattern='.tif$', full.names = TRUE)

# ECCC_CH ----
SPECIES <-  rast(ECCC_CH_LST)
# Reclassify into binary
terra::NAflag(SPECIES ) <- 128
BINARY <- pblapply(SPECIES , function(x) {
  x[x > 0] <- 1
})

# Get cumulative area
HA_SUM <- sum(SPECIES , na.rm = TRUE)
writeRaster(HA_SUM, "Data/Output/Richness/ECCC_CH_HA_SUM.tif", overwrite = TRUE)

# Get count of species
N <- sum(BINARY, na.rm = TRUE)
writeRaster(N, "Data/Output/Richness/ECCC_CH_N.tif", overwrite = TRUE)

# ECCC_CH ----
SPECIES <-  rast(ECCC_SAR_LST)
# Reclassify into binary
terra::NAflag(SPECIES) <- 128
BINARY <- pblapply(SPECIES , function(x) {
  x[x > 0] <- 1
})

# Get cumulative area
HA_SUM <- sum(SPECIES , na.rm = TRUE)
writeRaster(HA_SUM, "Data/Output/Richness/ECCC_SAR_HA_SUM.tif", overwrite = TRUE)

# Get count of species
N <- sum(BINARY, na.rm = TRUE)
writeRaster(N, "Data/Output/Richness/ECCC_SAR_N.tif", overwrite = TRUE)

## End timer
end_time <- Sys.time()
end_time - start_time

File: Python/species_richness.py
#
# Authors: Dan Wismer
#
# Date: June 28th, 2023
#
# Description: Generates species richness with option to subset by threat level. 
#              ie. a) sums up total area and b) sums up count 
#     
# Inputs:  1. Species rasters
#          2. Source name
#          3. Threat level
#          4. output FDGB
#
# Outputs: 1. Cumulative area ha raster
#          2. Count of species that intersect raster
#
#===============================================================================

import arcpy
from arcpy.sa import *

# Sara status 
def get_sara_status(code):
  if code == 0:
    return "" # NULL
  elif code == 1 or code == "Extirpated":
    return "EXT" # extirpated
  elif code == 2 or code == "Endangered":
    return "END" # endangered
  elif code == 3 or code == "Threatened":
    return "THR" # threatened
  elif code == 4 or code == "Special Concern":
    return "SPC" # special concern
  elif code == 5 or code == "No Status":
    return "NOS" # no status
  elif code == 6 or code == "Not at Risk":
    return "NAR" # not at risk

# Get user input
species = arcpy.GetParameterAsText(0).split(";")
folder = arcpy.GetParameterAsText(1)
source = arcpy.GetParameterAsText(2)
threat = arcpy.GetParameterAsText(3)

# Set threat code and wildcard
arcpy.AddMessage("Threat: {}".format(threat))
if threat != "All":
  threat = get_sara_status(threat)
  wildcard = "*{}*".format(threat)
else: 
  threat = "ALL"
  wildcard = "*"

# Get list of species .tifs if folder is provided
if folder:
  arcpy.env.workspace = folder
  species = arcpy.ListRasters(wildcard, "TIF")
  
# Mosaic - sum
arcpy.AddMessage(" ... Generating cumulative area. N = {}" .format(len(species)))
arcpy.MosaicToNewRaster_management(
  input_rasters = species,
  output_location = "Data/Output/Richness",
  raster_dataset_name_with_extension = "{}_{}_HA_SUM.tif".format(source, threat),
  pixel_type = "16_BIT_UNSIGNED",
  number_of_bands = 1,
  mosaic_method = "SUM"
)

# Reclass
arcpy.AddMessage("... Prepping rasters")
species_count = []
counter = 1
for x in species:
  arcpy.AddMessage("   ... {} ({}/{})".format(x, counter, len(species)))
  binary = Con(x, 1)
  species_count.append(binary)
  counter += 1
  
# Mosaic - count
arcpy.AddMessage("... Generating species count" .format(len(species)))
arcpy.MosaicToNewRaster_management(
  input_rasters = species_count, 
  output_location = "Data/Output/Richness", 
  raster_dataset_name_with_extension = "{}_{}_N.tif".format(source, threat),
  pixel_type = "16_BIT_UNSIGNED",
  number_of_bands = 1,
  mosaic_method = "SUM"
)
